### DISCUSSION TITLE
Beginner question: get genre from album of an given artist

### DISCUSSION THREAD

**Discussion by TheSud** [joalla/discogs_client#98] [Support] [answered]
Hi there!

Trying for weeks now to get the subject working in my tagging script. Didn't find anything useful via google or your documentation, because they all reference a release.id, which I obviously don't can fetch (correctly). Here are the relvant code snippets:
```
...
d = discogs_client.Client('my_user_agent/1.0', user_token='---snip---')
artists = d.search("Opeth", type='artist')
first_found_artist = artists.page(0)[0].data #the first found artist is working 99,7% of the time, so let's stick to it
release_dic = d.artist(first_found_artist['id']).releases
first_found_release = release_dic.page(0)[0].data #same applies to releases of the found artist above
try:
   release_details = d.release(first_found_release['id']) #PINPOINTED THE FAIL TO THIS LINE
   release_genre = release_details.genres
   genres_dic = d.release(first_found_release['id']).genres
   print ("Found", len(genres_dic), "genres:", genres_dic)
   first_found_genre = genres_dic[0]
                                
except Exception:
   first_found_genre = "Pop" #if nothing works/found, then use Pop as last resort, but we need a genre

try:
   styles_dic = d.release(first_found_release['id']).styles
   print ("Also found", len(styles_dic), "styles:", styles_dic)
                                
except Exception:
   pass
...
```

Expected Result: Opeth -> Album "Orchid" -> Genre "Rock" -> Style "Death Metal"
Got instead: Opeth -> Album "Frequencies" by LFO -> Genre "Electronic" -> Style "Techno"
because both albums share the same id (3649) and if I use the release-class LFO wins over Opeth. I used to query databases for years, so there must be a unique id somerwhere, which I can use - right? 

Any help with a beginner-friendly answer would be appreciated!

regards
Chris

**Comment by prcutler** [comment, marked as answer]
If I had to guess:

This query is giving you 3649 as the result, but it's the master ID, no the release ID for the Opeth album:
`first_found_release` (See: https://www.discogs.com/master/3649-Opeth-Orchid)

I don't have time to test the code right now, but can you update that query to get a result of a release ID and not a master ID?

Or can you add code to see if the result you're getting is equal to d.master.id - because that's not what you want, you want d.release_id, similar to what you have.  But does the first ['id] always return a master ID or release ID?

I hope that makes sense, I'm not an expert.  :)

**Reply by TheSud** [reply]
THX, looks like you gave me the right hint. But it seems to be the other way round: every album has an id (not a release id), but only some have a main_release (no master ID). Here's my output of the album in question:
`{'id': 3649, 'title': 'Orchid', 'type': 'master', 'main_release': 5076240, 'artist': 'Opeth', 'role': 'Main', 'resource_url': 'https://api.discogs.com/masters/3649', 'year': 1995, 'thumb': 'https://i.discogs.com/udGMWbW0NbH6wjkdgbuqNYKB8Lk2bSWApzt97_d6FRg/rs:fit/g:sm/q:40/h:150/w:150/czM6Ly9kaXNjb2dz/LWRhdGFiYXNlLWlt/YWdlcy9SLTUwNzYy/NDAtMTQ2MDk4NTIx/OC03NzUyLmpwZWc.jpeg', 'stats': {'community': {'in_wantlist': 276, 'in_collection': 606}, 'user': {'in_wantlist': 0, 'in_collection': 0}}}`

If I now use **d.release(first_found_release['main_release'])** in my code I get the correct album (at least on my test system with a few 100s of mp3-files), but also an exception if no main_release exists. So I check the existence of the main_release-key and use the ambiguous id if none is found as last resort for searching, because it's the only id I got. This seems to do the trick and I get correct results for these albums too.

Long explanation, short meaning: I will do extensive testing this weekend and report back on monday with the corrected code.

**Reply by TheSud** [reply]
As promised the corrected code, which worked for me over all my (main)releases:
```

                        d = discogs_client.Client('my_user_agent/1.0', user_token='---snip---')
                        artists = d.search("Opeth", type='artist')
                        first_found_artist = artists.page(0)[0].data
                        release_dic = d.artist(first_found_artist['id']).releases
                        first_found_release = release_dic.page(0)[0].data
                        
                        # First we try the 'lesser' unique id, then the better (but not always existing) main_release...
                        if "id" in first_found_release:
                            search_id = first_found_release['id']
                            
                        if "main_release" in first_found_release:
                            search_id = first_found_release['main_release']
                                      
                        try:
                            release_details = d.release(search_id)
                            genres_dic = release_details.genres
                                
                        except Exception:
                            # if no genre is found, we use the "allmighty" (but meaningless) Pop ;-)
                            print ("! EXCEPTION while using search_id...")
                            first_found_genre = "Pop"

                        try:
                            styles_dic = release_details.styles
                            print ("! DEBUG - Found", len(styles_dic), "styles:", styles_dic)
                                
                        except Exception:
                            pass


```

Thanx again for giving me the right hint - I mark the question as solved!

**Comment by prcutler** [comment]
Awesome, glad you were able to work it out.  If you have any more questions, please let us know!
